reproduce/face_rec.py

```diff
@@ -0,0 +1,106 @@
+    """
+
+    def __init__(self):
+        self.ref_img_path = None
+        self.known_faces = list()
+
+    def convert_bounding_boxes(self, bboxes):
+        """
+        Converts pipeline bounding box for use in face recognition
+        param bbox: bounding boxes in [left, top, width, height] form
+        :return: bboxes in [top, right, bottom, left]
+        """
+        faces = []
+        for bbox in bboxes:
+            left = bbox[0]
+            top = bbox[1]
+            w = bbox[2]
+            h = bbox[3]
+                        
+            right = w + left
+            bottom = top - h
+
+            faces.append([top, right, bottom, left])
+        return faces
+        
+
+    def get_ref_image(self, img_path):
+        """
+        If the user passes a reference image file path to the command line, it will be grabbed from here
+        param img_path: Path to the reference image to be used by face recognition
+        """
+        img = face_recognition.load_image_file(img_path)
+        self.known_faces.append(face_recognition.face_encodings(img)[0])
+
+    def generate_ref_image(self, bbox, frame):
+        """
+        Save a bounded location as a reference image
+
+        param bbox: The bounding box that contains the user's reference image
+        param frame: The frame that the bounding box is being used in
+        """
+        self.known_faces.append(
+            face_recognition.face_encodings(frame, known_face_locations=[bbox])[0]
+        )
+
+    def facerec_check(self, frame, device="cpu"):
+        face_locations = []
+        face_encodings = []
+
+
+        # Find all the faces and face encodings in the current frame of video
+        face_locations = face_recognition.face_locations(frame)
+        face_encodings = face_recognition.face_encodings(frame, face_locations)
+
+        for face_encoding in face_encodings:
+            # See if the face is a match for the known face(s), note that tolerance is low
+            match = face_recognition.compare_faces(
+                self.known_faces, face_encoding, tolerance=0.10
+            )
+
+            if match[0]:
+                #returns in [top, right, bottom, left] format
+                top = face_locations[0][0]
+                right = face_locations[0][1]
+                bottom = face_locations[0][2]
+                left = face_locations[0][3]
+                
+                h = top - bottom
+                w =  right - left
+                
+                #We want left, top, width, height
+                return (top, left, w, h)
+            else:
+                return None
+    
+    def select_face(self, bboxes, frame, tolerance=0.10):
+        """
+        selects a correct face from candidates bbox in frame
+        :param bboxes: the bounding boxes of candidates
+        :param frame: the frame
+        :return: the cropped face and its bbox data
+        """
+        
+        face = None
```

Comment:
what is face = None doing here?

Reply:
In the original method, it will return a None if no faces are matched with. If a face is found in the list of bounding boxes, it will overwrite that none value and return the correct bounding box. This way it fits into the original function of the method.

Reply:
doesn't matter too much, was just saying that the variable face is not used after assignment here since returning None instead of face at the end